Add renovate.json configuration

thepetk opened the pull request:
### What does this PR do?:
This PR adds the renovate.json configuration for the registry repo. The configuration will cover 2 things:

**Stack image updates**
The renovate bot will look inside the registry directories for `devfile.y(a)ml` files and try to update an image (looking for "image:") tag. For every different image found it will create a new PR. So if two different stacks use the same image, one PR will be created updating both image tags. On the other hand if we have 2 different images to update, it will create a PR for each one of them. Note, that images with "latest" tag will not be updated.

In order to remove the usage of variable values or latest (1 case) version, some tags have been updated to a fixed value. For nodeJS 16 the version was taken from all other stacks using the same image. On the first run, renovate should create a PR to update all of them.

**Github Actions**
Renovate will also look on every run, to update any deprecated github actions. All updates for github actions will be included on the same PR.

**Documentation**
The `contributing.md` guide was updated in order to emphasize to the usage of non-latest image tags when creating new stacks

**Next steps**
After merging this PR we need to follow all instructions [here](https://github.com/platform-engineering-org/renovate-runner#onboarding-your-githubcom-project) in order to add the registry project to the self hosted renovate bot.

### Which issue(s) this PR fixes:

fixes https://github.com/devfile/api/issues/1139

fixes https://github.com/devfile/api/issues/1138

### PR acceptance criteria:

- [x] Contributing guide
_Have you read the [devfile registry contributing guide](https://github.com/devfile/registry/blob/main/CONTRIBUTING.md) and followed its instructions?_
- [ ] Test automation
_Does this repository's tests pass with your changes?_
- [x] Documentation
_Does any documentation need to be updated with your changes?_
- [x] Check Tools Provider
_Have you tested the changes with existing tools, i.e. Odo, Che, Console? (See [devfile registry contributing guide](https://github.com/devfile/registry/blob/main/CONTRIBUTING.md) on how to test changes)_


### How to test changes / Special notes to the reviewer:

michael-valdron commented:
/retest

thepetk commented:
/retest

michael-valdron commented:
/retest

michael-valdron commented on CONTRIBUTING.md:
```diff
@@ -36,14 +36,16 @@ The following are required to build the devfile index container image containing
     - Make sure the name matches the devfile stack's name and be in the format `<language>-<framework>`. E.g. `java-quarkus`, `python-django`, etc.
 
-5) Add the devfile.yaml and any other necessary files for the stack under the stack folder.
+5) In case one of your components has an image, use a fixed version tag (e.g `<image>:1.1.0`) instead of `latest`. Our renovate bot will take care of the updates to your image's tag.
```
```suggestion
5) In case one of your components has a reference to an image, use a fixed version tag (e.g `<image>:1.1.0`) instead of `latest`. Our renovate bot will take care of the updates to the image tags.
```

michael-valdron commented on CONTRIBUTING.md:
```diff
@@ -59,6 +61,10 @@ Updating an existing devfile stack is relatively straightforward:
 
+### Automatic Stack Image Update
+
+As images used inside the stacks need to be up-to-date and in order to avoid using the `latest` tag, the renovate bot runs periodically ensuring that all images used from stacks (for example, inside components) have the latest version. As a result, all images used inside a devfile of a stack need to have a fixed version (e.g `1.1.0`) instead of `latest`.
```
```suggestion
Avoid using the `latest` tag, the renovate bot runs periodically ensuring that all image references used from stacks (for example, inside components) have the latest version. As a result, all image references used inside the stack devfiles will need to have a fixed version (e.g `1.1.0`) instead of `latest`.
```

michael-valdron reviewed (changes requested):
Aside from the feedback looks good! You also might need to contact the stack owners for approval on these stack changes.

cc: @BethGriggs @ajm01

thepetk commented:
> Aside from the feedback looks good! You also might need to contact the stack owners for approval on these stack changes.
> 
> cc: @BethGriggs @ajm01

@michael-valdron thanks for the review
I've updated the PR with the suggested changes and I've added @BethGriggs & @ajm01 as reviewers

yangcao77 commented on stacks/java-openliberty-gradle/devfile.yaml:
```diff
@@ -45,7 +45,7 @@ variables:
   - name: dev
     container:
-      image: icr.io/appcafe/open-liberty-devfile-stack:{{liberty-version}}-gradle
+      image: icr.io/appcafe/open-liberty-devfile-stack:22.0.0.1-gradle
```
should also remove the global variable

thepetk replied:
I've kept the global variable as it is still used on the stack in other places. I think upon reviewing the renovate bot PR, the reviewer should update also the value of the variable in order to get test passed (in case there is an update for this image)

yangcao77 commented:
@scottkurz @mezarin Also tagged the `java-liberty` stack owners to review

michael-valdron commented:
I have sent DMs to both @BethGriggs and @ajm01 under the [Kubernetes community slack workspace](https://kubernetes.slack.com), I would say give them an additional day or so to respond.

yangcao77 commented:
/lgtm

thepetk commented:
@ajm01 @mezarin and @scottkurz kind reminder for the review request of this PR :) Thanks a lot!

michael-valdron reviewed (changes requested):
We have not heard from @ajm01 yet, I recommend we for now revert the changes to the `java-openliberty-gradle`, `java-openliberty`, `java-websphereliberty-gradle`, and `java-websphereliberty` stacks and add an exclusion in renovate for these stacks.

We can open another PR to reintroduce these changes once we hear back. wdyt?

scottkurz commented:
Hi, I'm sorry this slipped past us and I'm just noticing this now.

The bigger issue here is that we have not been very current in updating the Liberty stack images, (the last one was 22.0.0.1).  Whether this change is made or not, our image is still almost two years old at this point.  Perhaps we need to consider abandoning our custom image and building a devfile based on a well-known Java/Maven image with a "tools"-type of component.

I'm getting ahead of myself.  Let's ignore that bigger issue though and just talk through the PR here:

If I understand it correctly, this might not be of too much value given the way the four Liberty stacks are maintained at the moment.   They use special-purpose Liberty images developed for devfile use cases (for purposes such as caching Maven dependencies in the image) rather than more general Liberty images.

So for the team updating these stack images, it's not too much more work to send a PR to this devfile registry repo at the same time.

That said, maybe it's still nicer to manage the versions getting updated via renovate and not having this extra PR to approve and merge. 

The only problem though is that the two Gradle stacks use the `liberty-version` variable in the run command: 

> commandLine: echo "gradle run command"; {{gradle-cmd}} -Dgradle.user.home=/.gradle libertyDev -Pliberty.runtime.version={{liberty-version}} -Pliberty.runtime.name=wlp-javaee8 -Pliberty.runtime.group=com.ibm.websphere.appserver.runtime --libertyDebug=false --hotTests --compileWait=3
  
So updating the image version without updating this version variable as well could lead to a confused experience.

Not sure where that leaves us.   Maybe we should just exempt the Liberty stacks?

thepetk commented:
> Hi, I'm sorry this slipped past us and I'm just noticing this now.
> 
> The bigger issue here is that we have not been very current in updating the Liberty stack images, (the last one was 22.0.0.1). Whether this change is made or not, our image is still almost two years old at this point. Perhaps we need to consider abandoning our custom image and building a devfile based on a well-known Java/Maven image with a "tools"-type of component.
> 
> I'm getting ahead of myself. Let's ignore that bigger issue though and just talk through the PR here:
> 
> If I understand it correctly, this might not be of too much value given the way the four Liberty stacks are maintained at the moment. They use special-purpose Liberty images developed for devfile use cases (for purposes such as caching Maven dependencies in the image) rather than more general Liberty images.
> 
> So for the team updating these stack images, it's not too much more work to send a PR to this devfile registry repo at the same time.
> 
> That said, maybe it's still nicer to manage the versions getting updated via renovate and not having this extra PR to approve and merge.
> 
> The only problem though is that the two Gradle stacks use the `liberty-version` variable in the run command:
> 
> > commandLine: echo "gradle run command"; {{gradle-cmd}} -Dgradle.user.home=/.gradle libertyDev -Pliberty.runtime.version={{liberty-version}} -Pliberty.runtime.name=wlp-javaee8 -Pliberty.runtime.group=com.ibm.websphere.appserver.runtime --libertyDebug=false --hotTests --compileWait=3
> 
> So updating the image version without updating this version variable as well could lead to a confused experience.
> 
> Not sure where that leaves us. Maybe we should just exempt the Liberty stacks?

Hey! No worries! I think the best options would be to exclude for now those stacks from the automation.

I think we can re-visit those stacks once we have more releases. This way we can have also the time to investigate a solution to use renovate with variables like `liberty-version`.

Just tfyi, with the current workaround the reviewer would be responsible to update the `liberty-version` variable value in case of a renovate update.

I'll come back shortly and exclude those stacks from this PR.

cc @michael-valdron  @scottkurz

thepetk commented:
@scottkurz & @michael-valdron I've reverted all updates to "liberty-related" stacks. I've also added a rule in renovate.json to ignore the paths of these stacks just to unnecessary error logging on the renovate runner.

michael-valdron reviewed (approved):
/lgtm
